src/libcore/intrinsics.rs

```diff
@@ -1917,6 +1917,15 @@ extern "rust-intrinsic" {
     #[rustc_const_unstable(feature = "const_discriminant", issue = "69821")]
     pub fn discriminant_value<T>(v: &T) -> <T as DiscriminantKind>::Discriminant;
 
+    /// Returns the number of variants of the type `T` cast to a `usize`;
```

Comment:
Returning a `usize` is incompatible with [`#![feature(repr128)]`](https://github.com/rust-lang/rust/issues/56071).

We don't currently expose a way for users to determine the discriminant type of an enum (see https://github.com/rust-lang/rust/pull/70705). I think we have a few options here:

1. Make this function return a `u128`. Assuming we never add `u256/i256` to the language (both of which seem insane as an enum discriminant type), this would be compatible with enums with any number of variants.
2. Stabilize the `DiscriminantKind1, and have this function return a value of type `<T as DiscriminantKind>::Discrimininant>`. This does not solve the problem of converting the value to an integer, but it avoids returning  a `u128` for enums which don't need it.
3. Declare that enums with more than `usize::MAX` *variants* are unsupported (`#[repr(u128)]` enums with fewer variants are fine). We might want to consider changing this to `u64`, though I don't actually see any use-case for having an enum with more than 4 billion variants (let alone on a 32-bit platform).

Reply:
On second thought, Rust does have tier-3 support for a 16-bit platform (AVR). While having `u16::MAX` variants seems ridiculous, it seems *slightly* more questionable to forbid that compared to `u32::MAX` (though again, I have no idea why anyone would do such a thing). Perhaps a `u32` is the way to go here.

Reply:
Seeing as the compiler uses an `IdxVec` to track declared variants we’re guaranteed to ICE if we have an enum with a number of variants exceeding `usize::MAX` anyway so I think they’re implicitly unsupported.

Reply:
Making it explicit is no bad thing though!

Reply:
Ah - I completely forgot about cross compilation... `u64` seems like the best bet then

Reply:
I personally believe `usize` is the correct type here. I do not want nor see us ever getting even close to compiling such an enum

Reply:
Ok... Thinking about this some more, I think we stick with `usize` and add a note to say that if your enum has more than `usize::MAX` variants on the target platform, the return value is unspecified. The probability this ever happens is minuscule and we've covered our backs with the note.

Reply:
> Seeing as the compiler uses an IdxVec to track declared variants we’re guaranteed to ICE if we have an enum with a number of variants exceeding usize::MAX anyway so I think they’re implicitly unsupported.

The compiler could be running on a 64bit system but generate code for a 16bit system. So you cannot really deduce anything from the type the compiler is using here.

Reply:
I’m in favour of `<T as DiscriminantKind>::Discrimininant>` or `u128` or `u64`, in that specific order. Having safe and stable APIs that may return unspecified or incorrect results without a very very convincing reason is questionable.

Reply:
Unfortunately, you can’t use `<T as DiscriminantKind>::Discriminant` because if I have an enum with `256` variants the discriminant type will be `u8` which can’t represent `256`.

Reply:
> Seeing as the compiler uses an `IdxVec` to track declared variants we’re guaranteed to ICE if we have an enum with a number of variants exceeding `usize::MAX` anyway so I think they’re implicitly unsupported.

I should have struck this out because it doesn’t apply anyway in the case of cross comp